lib/compress.js

```diff
@@ -479,7 +479,7 @@ merge(Compressor.prototype, {
                     if (ref.scope.uses_eval || ref.scope.uses_with) break;
 
                     // Constant single use vars can be replaced in any scope.
-                    if (var_decl.value.is_constant()) {
+                    if (!(stat instanceof AST_ForIn) && var_decl.value.is_constant()) {
                         var ctt = new TreeTransformer(function(node) {
                             if (node === ref) {
                                 var parent = ctt.parent();
```

Comment:
I've changed this patch a few times. I'm not happy with it. Might have a potential issue on `harmony` with pathological ES6 input. I'll revisit it tomorrow.